### DISCUSSION TITLE
How to create a Bitmap from IntPtr?

### DISCUSSION THREAD

**Discussion by suoyukii** [AvaloniaUI/Avalonia#16703] [Q&A] [answered]
I used the Win32 API's PrivateExtractIcons to get an IntPtr of an icon, how can I convert this IntPtr into a Bitmap for AvaloniaUI?

```csharp
    [DllImport("user32")]
    private static extern int PrivateExtractIcons(string lpsz_file, int n_icon_index, int cx_icon, int cy_icon,
        IntPtr[] phicon, int[] piconid, int n_icons, int flags);

    [DllImport("user32")]
    private static extern void DestroyIcon(nint h_icon);
```
```csharp
            var file_url = @"C:\Windows\System32\user32.dll";
            var icon_total_count = PrivateExtractIcons(file_url, 0, 0, 0, null, null, 0, 0);
            var h_icons = new IntPtr[icon_total_count];
            var success_count = PrivateExtractIcons(file_url, 0, 32, 32, h_icons, new uint[icon_total_count],
                icon_total_count, 0);
            if (success_count <= 0) return;

            foreach (var ptr in h_icons)
            {
                if (ptr == IntPtr.Zero) continue;
                // The variable ptr is IntPtr.
                Content = new Image
                {
                    // Source = new Bitmap()
                    // Here is the method to convert the IntPtr of wpf to BitMap.
                    // Imaging.CreateBitmapSourceFromHIcon(ptr, Int32Rect.Empty, BitmapSizeOptions.FromEmptyOptions())
                };
                break;
            }

            foreach (var ptr in h_icons) DestroyIcon(ptr);
```

**Comment by stevensrmiller** [comment]
Could you post the code you are using to get the IntPtr?

**Reply by suoyukii** [reply]
Code has been provided.

**Reply by suoyukii** [reply]
Is there really no way to convert IntPtr to Bitmap in AvanloniaUI?

**Reply by thevortexcloud** [reply]
That really depends on what the pointer points to. But I am also not that familiar with the win32 icon APIs to say if there is a specific way to do it. 

But there is no direct way to do it that I know of. Assuming you can get that pointer into a byte array of the the actual image data, then it should be very easy to do. 

Looking at the source code for the WPF method seems to lead to this method being called:

https://learn.microsoft.com/en-us/windows/win32/api/wincodec/nf-wincodec-iwicimagingfactory-createbitmapfromhicon

Which looks like it returns something that has some methods that can be used to get the data into a byte array.

**Reply by maxkatz6** [reply]
> Is there really no way to convert IntPtr to Bitmap in AvanloniaUI?

It's easy to convert IntPtr pointing to the pixels buffer to the Bitmap. See this method:
https://github.com/AvaloniaUI/Avalonia/blob/45a2d36e3214ac4b3623940ba728b2503de5c668/src/Avalonia.Base/Media/Imaging/Bitmap.cs#L104
Obviously, you need to know all of these parameters of your pixel buffer.
When integrating with third party libraries, you usually know (or can assume) most of these parameters.

But here your IntPtr points to the HICON instance.
HICON **is not** a pixel buffer. It's a complex Windows memory strucutre.
And you need to use other Windows APIs first to get a pixel buffer from it.

**Reply by maxkatz6** [reply]
@suoyukii answer from @stevensrmiller below actually shows one of possible ways of getting pixel buffer from the HIcon using `System.Drawing.Common` helpers.

**Reply by stevensrmiller** [reply]
In the Win32 API, any type that starts with an "H," or is called a "handle," isn't a pointer. It's an opaque reference. You can't do anything with its value, even if you could find out what it was, because it isn't an address. It's an identifier used to look up a data structure Windows maintains internally. In some ways, it's similar to an object reference in a managed language like C# or Java.

To get the data a handle refers to, you pretty much always need to use that handle as an argument to some system API call.

**Comment by stevensrmiller** [comment]
This is not very pretty, but it might do what you want:

    using System.Drawing;
    using System.Drawing.Imaging;
    using Avalonia;
    using Avalonia.Controls;
    using aBrushes = Avalonia.Media.Brushes;
    using aBitmap = Avalonia.Media.Imaging.Bitmap;
    
    class Project
    {
        public static void Main(string[] args)
        {
            AppBuilder.Configure<Application>()
                      .UsePlatformDetect()
                      .Start(AppMain, args);
        }
    
        public static void AppMain(Application app, string[] args)
        {
            app.Styles.Add(new Avalonia.Themes.Fluent.FluentTheme());
            app.RequestedThemeVariant = Avalonia.Styling.ThemeVariant.Light;
    
            var sp = new StackPanel
            {
                Background = aBrushes.Black,
            };
    
            for (int n = 0; n < 7; ++n)
            {
                Icon icon =
                    Icon.ExtractIcon(@"C:\Windows\System32\user32.dll", n, true);
    
                //Icon.ExtractAssociatedIcon(@"C:\WINDOWS\system32\notepad.exe");
    
                Bitmap bmp = icon.ToBitmap();
        
                Rectangle rect = new Rectangle(0, 0, bmp.Width, bmp.Height);
        
                BitmapData  bitmapdata = bmp.LockBits
                (
                    rect,
                    ImageLockMode.ReadWrite,
                    bmp.PixelFormat
                );
    
                aBitmap bitmap1 = new aBitmap
                (
                    Avalonia.Platform.PixelFormat.Bgra8888,
                    Avalonia.Platform.AlphaFormat.Unpremul,
                    bitmapdata.Scan0,
                    new PixelSize(bitmapdata.Width, bitmapdata.Height),
                    new Vector(96, 96),
                    bitmapdata.Stride
                );
                
                bmp.UnlockBits(bitmapdata);
                bmp.Dispose();
    
                var image = new Avalonia.Controls.Image
                {
                    Source = bitmap1,
                    Height = bitmapdata.Height,
                    Width = bitmapdata.Width,
                };
    
                sp.Children.Add(image);
            }
    
            var win = new Window
            {
                Title = "An Icon",
                Content = sp,
                Background = aBrushes.Black,
            };
    
            win.Show();
            app.Run(win);
        }
    }

Here's the output I get:

![ScreenCap](https://github.com/user-attachments/assets/cdb4f8a8-d6f6-419d-935e-60c5274241bf)

Here's my .csproj:

    <Project Sdk="Microsoft.NET.Sdk">
    
      <PropertyGroup>
        <TargetFramework>net8.0</TargetFramework>
        <OutputType>Exe</OutputType>
      </PropertyGroup>
    
      <ItemGroup>
        <PackageReference Include="System.Drawing.Common" Version="8.0.8" />
        <PackageReference Include="Avalonia" Version="11.1.0" />
        <PackageReference Include="Avalonia.Desktop" Version="11.1.0" />
        <PackageReference Include="Avalonia.Themes.Fluent" Version="11.1.0" />
      </ItemGroup>
    
    </Project>

That's all based on something I found here: https://github.com/AvaloniaUI/Avalonia/discussions/5908

**Reply by suoyukii** [reply, marked as answer]
# .csproj
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net8.0</TargetFramework>
    <OutputType>Exe</OutputType>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="System.Drawing.Common" Version="8.0.8" />
    <PackageReference Include="Avalonia" Version="11.1.0" />
    <PackageReference Include="Avalonia.Desktop" Version="11.1.0" />
    <PackageReference Include="Avalonia.Themes.Fluent" Version="11.1.0" />
  </ItemGroup>

</Project>
```

```csharp
using System.Drawing;
using System.Drawing.Imaging;
using System.IO;
using System.Runtime.InteropServices;
using Avalonia;
using Avalonia.Controls;
using Avalonia.Input;
using Avalonia.Platform;
using AvaloniaImage = Avalonia.Controls.Image;
using AvaloniaBitmap = Avalonia.Media.Imaging.Bitmap;
```

# Code
```csharp
    [DllImport("user32")]
    private static extern int PrivateExtractIcons(string lpsz_file, int n_icon_index, int cx_icon, int cy_icon,
        nint[] phicon, uint[] piconid, int n_icons, int flags);

    [DllImport("user32")]
    private static extern void DestroyIcon(nint h_icon);
```

```csharp
            var file_url = @"C:\Windows\System32\shell32.dll";
            var size = 256;
            var icon_total_count = PrivateExtractIcons(file_url, 0, size, size, null, null, 0, 0);
            var h_icons = new nint[icon_total_count];
            var success_count = PrivateExtractIcons(file_url, 0, size, size, h_icons, new uint[icon_total_count],
                icon_total_count, 0);
            if (success_count <= 0) return;
            foreach (var ptr in h_icons)
            {
                if (ptr == nint.Zero) continue;
                var icon = System.Drawing.Icon.FromHandle(ptr);
                var bmp = icon.ToBitmap();
                icon.Dispose();
                DestroyIcon(ptr);

                var bmpdata = bmp.LockBits
                (
                    new Rectangle(0, 0, bmp.Width, bmp.Height),
                    ImageLockMode.ReadWrite,
                    bmp.PixelFormat
                );
                Content = new AvaloniaImage
                {
                    Width = 96,
                    Height = 96,
                    Margin = new Thickness(8),
                    Source = new AvaloniaBitmap(
                        Avalonia.Platform.PixelFormat.Bgra8888,
                        AlphaFormat.Unpremul,
                        bmpdata.Scan0,
                        new PixelSize(bmpdata.Width, bmpdata.Height),
                        new Vector(96, 96),
                        bmpdata.Stride
                    )
                };
                break;
            }
```

**Reply by suoyukii** [reply]
Thank you, I found that changing `System.Drawing.Icon.ExtractIcon` to `System.Drawing.Icon.FromHandle` allows me to read the data in the IntPtr.

**Reply by stevensrmiller** [reply]
Yes, it does. But you have to convert it and then destroy it, and still use WIn32 calls with DLL imports. That's not really necessary.

**Reply by suoyukii** [reply]
The win32 api is called to get the total number of icons for a given resolution.

**Reply by stevensrmiller** [reply]
Icon.ExtractIcon will return null when there's no icon at your index.
